File: scripts/pack_emojis.py
import json
import struct
import os

def pack_emojis():
    json_path = "data/emoji.json"
    bin_path = "src/emojis.bin"

    print(f"Reading {json_path}...")
    with open(json_path, "r", encoding="utf-8") as f:
        emojis_data = json.load(f)

    string_table = bytearray()
    string_offsets = {}

    def get_or_add_string(s):
        s_bytes = s.encode("utf-8") + b"\x00"
        if s_bytes in string_offsets:
            return string_offsets[s_bytes]
        
        offset = len(string_table)
        string_table.extend(s_bytes)
        string_offsets[s_bytes] = offset
        return offset

    entries = []
    
    for item in emojis_data:
        emoji_char = item.get("emoji")
        description = item.get("description", "")
        tags = item.get("tags", [])
        aliases = item.get("aliases", [])

        if not emoji_char:
            continue

        # Build a search string containing unique words in lowercase
        search_words = []
        for word in description.lower().replace("-", " ").replace("_", " ").split():
            clean_word = "".join(c for c in word if c.isalnum())
            if clean_word and clean_word not in search_words:
                search_words.append(clean_word)
        
        for tag_or_alias in tags + aliases:
            for word in tag_or_alias.lower().replace("-", " ").replace("_", " ").split():
                clean_word = "".join(c for c in word if c.isalnum())
                if clean_word and clean_word not in search_words:
                    search_words.append(clean_word)

        search_str = " ".join(search_words)

        emoji_offset = get_or_add_string(emoji_char)
        name_offset = get_or_add_string(description)
        search_offset = get_or_add_string(search_str)

        entries.append({
            "emoji_offset": emoji_offset,
            "name_offset": name_offset,
            "search_offset": search_offset
        })

    emoji_count = len(entries)
    print(f"Packed {emoji_count} emojis.")

    index_size = emoji_count * 12
    header_size = 16
    
    string_table_offset = header_size + index_size

    header = struct.pack(
        "<4sHHII",
        b"EMJG",
        1,
        emoji_count,
        string_table_offset,
        len(string_table)
    )

    index_data = bytearray()
    for entry in entries:
        index_data.extend(struct.pack(
            "<III",
            entry["emoji_offset"],
            entry["name_offset"],
            entry["search_offset"]
        ))

    os.makedirs(os.path.dirname(bin_path), exist_ok=True)
    with open(bin_path, "wb") as f:
        f.write(header)
        f.write(index_data)
        f.write(string_table)

    total_size = len(header) + len(index_data) + len(string_table)
    print(f"Binary generated at {bin_path} ({total_size / 1024:.2f} KB).")

if __name__ == "__main__":
    pack_emojis()

File: scripts/pack_emojis.go
// SPDX-FileCopyrightText: 2026 Uwe Jugel
// SPDX-License-Identifier: AGPL-3.0-or-later

package main

import (
	"bytes"
	"encoding/binary"
	"encoding/json"
	"fmt"
	"os"
	"path/filepath"
	"strings"
)

type EmojiItem struct {
	Emoji       string   `json:"emoji"`
	Description string   `json:"description"`
	Tags        []string `json:"tags"`
	Aliases     []string `json:"aliases"`
}

func cleanWord(word string) string {
	var sb strings.Builder
	for _, r := range word {
		if (r >= 'a' && r <= 'z') || (r >= 'A' && r <= 'Z') || (r >= '0' && r <= '9') {
			sb.WriteRune(r)
		}
	}
	return sb.String()
}

func main() {
	jsonPath := "data/emoji.json"
	binPath := "src/emojis.bin"

	fmt.Printf("Reading %s...\n", jsonPath)
	data, err := os.ReadFile(jsonPath)
	if err != nil {
		fmt.Printf("Error reading JSON file: %v\n", err)
		os.Exit(1)
	}

	var emojis []EmojiItem
	if err := json.Unmarshal(data, &emojis); err != nil {
		fmt.Printf("Error unmarshaling JSON: %v\n", err)
		os.Exit(1)
	}

	stringTable := &bytes.Buffer{}
	stringOffsets := make(map[string]uint32)

	getOrAddString := func(s string) uint32 {
		sBytes := append([]byte(s), 0) // Null terminator
		sKey := string(sBytes)
		if off, exists := stringOffsets[sKey]; exists {
			return off
		}
		off := uint32(stringTable.Len())
		stringTable.Write(sBytes)
		stringOffsets[sKey] = off
		return off
	}

	type Entry struct {
		EmojiOff  uint32
		NameOff   uint32
		SearchOff uint32
	}

	var entries []Entry

	for _, item := range emojis {
		if item.Emoji == "" {
			continue
		}

		// Collect unique search words
		var searchWords []string
		addWords := func(text string) {
			cleaned := strings.ReplaceAll(text, "-", " ")
			cleaned = strings.ReplaceAll(cleaned, "_", " ")
			for _, w := range strings.Fields(strings.ToLower(cleaned)) {
				cw := cleanWord(w)
				if cw == "" {
					continue
				}
				// check uniqueness
				found := false
				for _, sw := range searchWords {
					if sw == cw {
						found = true
						break
					}
				}
				if !found {
					searchWords = append(searchWords, cw)
				}
			}
		}

		addWords(item.Description)
		for _, t := range item.Tags {
			addWords(t)
		}
		for _, a := range item.Aliases {
			addWords(a)
		}

		searchStr := strings.Join(searchWords, " ")

		cleanEmoji := strings.ReplaceAll(item.Emoji, "\uFE0F", "")
		emojiOff := getOrAddString(cleanEmoji)
		nameOff := getOrAddString(item.Description)
		searchOff := getOrAddString(searchStr)

		entries = append(entries, Entry{
			EmojiOff:  emojiOff,
			NameOff:   nameOff,
			SearchOff: searchOff,
		})
	}

	emojiCount := uint16(len(entries))
	fmt.Printf("Packed %d emojis.\n", emojiCount)

	// Index size: count * 12 bytes
	indexSize := uint32(emojiCount) * 12
	headerSize := uint32(16)
	stringTableOffset := headerSize + indexSize

	// Write binary file
	outBuf := &bytes.Buffer{}

	// Header: magic (4s), version (u16), count (u16), string_table_offset (u32), string_table_len (u32)
	outBuf.Write([]byte("EMJG"))
	binary.Write(outBuf, binary.LittleEndian, uint16(1))
	binary.Write(outBuf, binary.LittleEndian, emojiCount)
	binary.Write(outBuf, binary.LittleEndian, stringTableOffset)
	binary.Write(outBuf, binary.LittleEndian, uint32(stringTable.Len()))

	// Index Entries
	for _, entry := range entries {
		binary.Write(outBuf, binary.LittleEndian, entry.EmojiOff)
		binary.Write(outBuf, binary.LittleEndian, entry.NameOff)
		binary.Write(outBuf, binary.LittleEndian, entry.SearchOff)
	}

	// String Table
	outBuf.Write(stringTable.Bytes())

	if err := os.MkdirAll(filepath.Dir(binPath), 0755); err != nil {
		fmt.Printf("Error creating directory: %v\n", err)
		os.Exit(1)
	}

	if err := os.WriteFile(binPath, outBuf.Bytes(), 0644); err != nil {
		fmt.Printf("Error writing binary file: %v\n", err)
		os.Exit(1)
	}

	fmt.Printf("Binary generated at %s (%.2f KB).\n", binPath, float64(outBuf.Len())/1024.0)
}
